### PR TITLE
SPI: Add Xilinx AXI Quad SPI (QSPI) driver

### PR DESCRIPTION
**Pull request by henrikbrixandersen**
Add driver for the Xilinx AXI Quad SPI v3.2 IP and enable it on the ARM Cortex-M1 DesignStart FPGA reference implementation running on the Digilent Arty.

### CODE REVIEW

**Comment by henrikbrixandersen** [comment]
@pabigot The `jesd216` sample added in 506f02e89ec94fde15eb578a8c835c01839365d9 fails on this board, since the output does not match the regex from `sample.yaml`:

```
*** Booting Zephyr OS build zephyr-v2.3.0-1907-g471e7969205d  ***
FLASH_0: SFDP v 1.0 AP ff with 0 PH
jedec-id = [20 ba 18];
```

The expected `sfdp-bfp =` line is missing. Please advice.

**Comment by pabigot** [comment]
> The expected `sfdp-bfp =` line is missing.

Bug in the sample; see #27629.

**Comment by henrikbrixandersen** [comment]
> Bug in the sample; see #27629.

@pabigot Thanks. With the fixes from #27629, the sample output is:

```
*** Booting Zephyr OS build zephyr-v2.3.0-1907-g471e7969205d  ***
FLASH_0: SFDP v 1.0 AP ff with 1 PH
PH0: ff00 rev 1.0: 9 DW @ 30
Summary of BFP content:
DTR Clocking not supported
Addressing: 3-Byte only
4-KiBy erase: uniform
Support 1-1-1
Support 1-1-2: instr 3Bh, 0 mode clocks, 8 waits
Support 1-1-4: instr 6Bh, 1 mode clocks, 7 waits
Support 1-2-2: instr BBh, 1 mode clocks, 7 waits
Support 1-4-4: instr EBh, 1 mode clocks, 9 waits
Support 2-2-2: instr BBh, 1 mode clocks, 7 waits
Support 4-4-4: instr EBh, 1 mode clocks, 9 waits
Flash density: 16777216 bytes
ET1: instr 20h for 4096 By
ET2: instr D8h for 65536 By
size = <134217728>;
sfdp-bfp = [
	e5 20 f1 ff  ff ff ff 07  29 eb 27 6b  08 3b 27 bb
	ff ff ff ff  ff ff 27 bb  ff ff 29 eb  0c 20 10 d8
	00 00 00 00
	];
jedec-id = [20 ba 18];
```

**Review by MaureenHelm — APPROVED** [review]

File: boards/arm/arty/arty_a7_arm_designstart_m1.dts [hunk 1]
```diff
@@ -150,6 +150,48 @@
 			reg = <0x20000000 DT_SIZE_K(32)>;
 		};
 
+		daplink_gpio0: gpio@40010000 {
```
Comment (pabigot) [inline boards/arm/arty/arty_a7_arm_designstart_m1.dts:153]: Is this something that should have been added when GPIO was added?  It looks like it belongs in a separate commit.

**Review by pabigot — CHANGES_REQUESTED** [review]
Within the limits of not being able to test this looks good.  I'm thinking the GPIO devicetree node belongs in a separate commit from the QSPI ones, though.

File: boards/arm/arty/arty_a7_arm_designstart_m1.dts (hunk 1 above)
Comment (henrikbrixandersen) [reply boards/arm/arty/arty_a7_arm_designstart_m1.dts]: @pabigot Thank you for reviewing! The `daplink_gpio0` controller is used for multiplexing access to the off-board QSPI SPI NOR flash (support for this will follow in a separate PR) between read-only XIP access and `jedec,spi-nor` driver access.

I have removed the GPIO node for now and will (re-)introduce it along with support for the MUX.

**Review by pabigot — APPROVED** [review]

File: drivers/spi/spi_xlnx_axi_quadspi.c [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+	struct xlnx_quadspi_data *data = dev->data;
+	struct spi_context *ctx = &data->ctx;
+	uint32_t temp;
+	uint32_t drr;
+
+	/* Acknowledge interrupt */
+	temp = xlnx_quadspi_read32(dev, IPISR_OFFSET);
+	xlnx_quadspi_write32(dev, temp, IPISR_OFFSET);
+
+	if (temp & IPIXR_DTR_EMPTY) {
```
Comment (tbursztyka) [inline drivers/spi/spi_xlnx_axi_quadspi.c:397]: I would do the inverse test here, print the message found in line 432 and return.

Would be more readable

**Review by tbursztyka — APPROVED** [review]
Just one tiny comment. But it looks proper already to be merged so approving.
